streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -1417,7 +1417,17 @@ class ZStream[-R, +E, +A](val channel: ZChannel[R, Any, Any, Any, E, Chunk[A], A
    * new elements.
    */
   final def mapAccumM[R1 <: R, E1 >: E, S, A1](s: S)(f: (S, A) => ZIO[R1, E1, (S, A1)]): ZStream[R1, E1, A1] =
-    ???
+    ZStream.unwrap(
+      for {
+        state <- Ref.make(s)
```

Comment:
We can now implement combinators using recursion on channels to avoid the Ref overhead. Check out how the other combinators are implemented. Let me know if you need a pointer and I will write one up when I'm on my computer.

Reply:
Hmm could you point to the one you mean? I tried refactoring based on the `mapAccum` above it, but hit a snag because it loses laziness to naively change to `ZChannel.fromEffect(chunk.mapAccumM(s)(f))`. Do you think I should still try to incorporate the `loopPartial...`?